lib/pbench/agent/tool_meister.py

```diff
@@ -1357,25 +1356,32 @@ def _send_directory(self, directory, uri, ctx):
         tar_file = parent_dir / f"{target_dir}.tar.xz"
 
         try:
-            cp = self._create_tar(directory, tar_file, retry=True)
-            if cp.returncode != 0:
-                self.logger.warning("Failed to create tarball, %s", cp.stdout)
+            if self._create_tar(directory, tar_file).returncode != 0:
                 # Tar ball creation failed even after suppressing all the warnings,
                 # we will now proceed to create an empty tar ball.
                 # TODO: it'd be better to be able to skip the PUT entirely if the
                 # tar fails and simply log a failure without TDS waiting forever.
-                if self._create_tar(Path("/dev/null"), tar_file) != 0:
+                if self._create_tar(Path("/dev/null"), tar_file).returncode != 0:
                     # Empty tarball creation failed, so we're going to skip the PUT
-                    # operation, and that's going to cause the TDS to hang.
-                    raise ToolMeisterError("Failed to create an empty tar")
+                    # operation.
+                    raise ToolMeisterError(
+                        f"Failed to create an empty tar {str(tar_file)}"
+                    )
+        except ToolMeisterError:
+            raise
         except Exception:
-            self.logger.exception("Failed to create tar ball, '%s'", tar_file)
+            self.logger.exception(
+                "Exception attempting to create the tarball, '%s'", tar_file
+            )
             failures += 1
```

Comment:
Weird that I can't seem to locate this comment in the changes tab to reply as part of a review. Rebasing does weird things to GitHub.

Anyway, the point was that he's raising `ToolMeisterError` inside this try block, and wants it to propagate rather than stopping it here. This is one of the problems with blindly finalizing overly-broad exceptions classes like `Exception` ... but at the time the hack seemed reasonably expedient vs trying to work out a real solution. 😬

Reply:
> Rebasing does weird things to GitHub.

Oh, it's even weirder than that (e.g., Nikhil has not yet rebased since I made the comment):  I started my review against only the last six commits, and ended it with comments against all 12, so if you're looking at the wrong set, you won't see my comment....  (And, for the record, the commit-range selector has been broken now for about 3 weeks...I need to report that...again.)  However, if you click on the [`lib/pbench/agent/tool_meister.py`](https://github.com/distributed-system-analysis/pbench/pull/2704/files/e260e6455d5a18e63e542c25b6bf195453102f19..9dc1f534e03403bc64b73e0bb4b845eb3e897496#diff-39405c6c2d5e89ba6fd6ba87a77fa9175dbdd28062af27e6d7c36f583bd0ba4b) link above the diffs above, you'll get there.  🙃 

> the point was that he's raising `ToolMeisterError` inside this try block, and wants it to propagate rather than stopping it here.

Yeah, I noted that, but I'm not clear on what that's accomplishing (as well as how the events leading up to the `ToolMeisterError` are any different from any other sort of exception-raising failure).

> This is one of the problems with blindly finalizing overly-broad exceptions classes like `Exception`

I noted that this scope is enclosed by a `try` block with a `finally` clause, and those are really resilient!  And, this one is prbably going to issue a warning when it cannot remove the tarball that we didn't create....

Reply:
> However, if you click on the [...] link above the diffs above, you'll get there.

This is actually true, and it takes me to the *6 commits* you had selected at the time. Since the comment did not appear on *any* individual commit, apparently GitHub manages comments made against a multiple-select set of commits *separately* so that they can only be found in conversation view or against that specific set.

This is a bit mind blowing and surprisingly bizarre. How to go out of your way to be opaque and mysterious, GitHub!

Reply:
I assume their thinking is that the comment relates to the specific differences being displayed, so displaying it for any other combination of commits might not make sense.  🦐

Reply:
@webbnh Yes I do think handling `ToolMeisterError` differently than the broad `exception` is worthwhile, because if this exception is raised then we are certain that we failed to even create an empty tarball and it is better to propagate this error to higher level rather than logging it locally and continuing with our operation. There is no point of moving forwards if we encounter this error in this place of our TM workflow.

Reply:
Looking at the code, it seems like the only difference between the handling of `ToolMeisterError` and any other `Exception` is that the later results in a normal return with a non-zero value while the former results in an exception propagation -- they both skip the rest of the normal processing code in `_send_directory()`, and they both execute the `finally` block.  The main difference occurs in the _callers_ of `_send_directory()` which look woefully unprepared to handle exceptions....

Nevertheless, if @dbutenhof is OK with this state of things, I shan't argue.